tests/std/tests/P1222R4_flat_set/test.cpp

```diff
@@ -235,6 +236,122 @@ void test_insert_2() {
     }
 }
 
+struct key_comparer {
+    const auto& extract_key(const auto& obj) const {
+        if constexpr (requires { obj.key; }) {
+            return obj.key;
+        } else {
+            return obj;
+        }
+    }
+
+    bool operator()(const auto& lhs, const auto& rhs) const {
+        return extract_key(lhs) < extract_key(rhs);
+    }
+
+    using is_transparent = int;
+};
+
+void test_comparer_application() {
+    // The set must rely on its comparer to do the comparisons.
+    struct incomparable {
+        int key;
+        bool operator<(const incomparable&) const  = delete;
+        bool operator==(const incomparable&) const = delete;
+    };
+
+    flat_set<incomparable, key_comparer> fs{{0}, {3}, {1}, {0}, {5}};
+    assert(fs.contains(0));
+    assert(!fs.contains(2));
+    fs.insert(fs.begin(), incomparable{4});
+    fs.insert(2);
+    assert(fs.contains(4));
+    assert(fs.contains(incomparable{2}));
+
+    assert(fs.lower_bound(3) == fs.lower_bound(incomparable{3}));
+    fs.erase(2);
+    assert(!fs.contains(incomparable{2}));
+}
+
+void test_insert_transparent() {
+    // For flat_set::insert([hint,]auto&&), the input should be unchanged if the set already
+    // contains an equivalent element.
+    struct detect_conversion {
+        int key;
+        mutable bool converted = false;
+
+        explicit operator int() const {
+            converted = true;
+            return key;
+        }
+    };
+
+    flat_set<int, key_comparer> fs{0, 3, 5};
+    assert_all_requirements_and_equals(fs, {0, 3, 5});
+    detect_conversion detector{3};
+
+    assert(!detector.converted);
+    fs.insert(detector /*3*/);
+    assert_all_requirements_and_equals(fs, {0, 3, 5});
+    assert(!detector.converted);
+
+    detector.key = 1;
+
+    assert(!detector.converted);
+    fs.insert(detector /*1*/);
+    assert_all_requirements_and_equals(fs, {0, 1, 3, 5});
+    assert(detector.converted);
+
+    detector.converted = false;
+
+    assert(!detector.converted);
+    fs.insert(fs.end(), detector /*1*/);
+    assert_all_requirements_and_equals(fs, {0, 1, 3, 5});
+    assert(!detector.converted);
+
+    detector.key = 2;
+
+    assert(!detector.converted);
+    fs.insert(fs.begin(), detector /*2*/);
+    assert_all_requirements_and_equals(fs, {0, 1, 2, 3, 5});
+    assert(detector.converted);
+}
+
+void test_insert_using_invalid_hint() {
+    mt19937 eng(42);
+
+    uniform_int_distribution<int> dist_seq(0, 20);
+
+    vector<int> seq(200);
+    for (int& val : seq) {
+        val = dist_seq(eng);
+    }
+
+    {
+        flat_multiset<int> with_hint, no_hint;
+        for (const int val : seq) {
+            uniform_int_distribution<int> dist_idx(0, int(with_hint.size()));

```

Comment:
`int(with_hint.size())` is a functional-style C-semantics cast, which we strongly avoid. Occurs below.